core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -869,7 +876,20 @@ private[spark] class BlockManager(
           blockInfoManager.unlock(blockId)
         }
       } else {
-        blockInfoManager.removeBlock(blockId)
+        if (exceptionWasThrown) {
```

Comment:
could also combine this with the above else

btw, is it necessary to unlock the block in this path?